### PR TITLE
feat(datastore): Custom Error Handler

### PR DESCRIPTION
**Pull request by fjnoyp**
*Issue #, if available:*

*Description of changes:*

Custom ErrorHandler support 

I haven't written integration tests for this.  Should be a small code change but a bit tricky to figure out as I don't know how to trigger the Datastore Exceptions that would in turn trigger the custom error handler (I've just manually triggered it from native atm) 

Update November 8: 

We should include integration tests for this code change in a separate PR.  Writing integration tests will require enabling cloud sync and thus reworking all existing integration tests in Datastore as well.  Given ongoing talks with native teams, I might also need to enable Auth as well to trigger the edge case that would get the custom error callback to be called in the first place.  As this could require significantly more time and is outside scope of original task I plan to tackle this in a separate line of work.  There are already unit tests for this feature though. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 1]
```diff
@@ -163,6 +165,18 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
 
         val dataStoreConfigurationBuilder = DataStoreConfiguration.builder()
 
+        if(request["hasCustomErrorHandler"] as Boolean) {
```
Comment (fjnoyp) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:168]: I think this naming `hasCustomErrorHandler` is inconsistent and should just be `hasErrorHandler`

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 2]
```diff
@@ -15,6 +15,7 @@
 
 package com.amazonaws.amplify.amplify_datastore
 
+import android.database.DefaultDatabaseErrorHandler
```
Comment (HuiSF) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:18]: Accidentally imported?

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 3]
```diff
@@ -163,6 +165,18 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
...
+        if( (request["hasErrorHandler"] as? Boolean? == true) ) {
+            var handler = DataStoreErrorHandler {
+                val args = hashMapOf(
+                        "errorCode" to "DataStoreException",
+                        "errorMessage" to ExceptionMessages.defaultFallbackExceptionMessage,
+                        "details" to createSerializedError(it)
+                )
+                channel.invokeMethod("errorHandler", args)
+            }
+            dataStoreConfigurationBuilder.errorHandler(handler!!)
```
Comment (HuiSF) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:177]: `!!` is probably unnecessary.

File: packages/amplify_datastore/example/lib/main.dart [hunk 4]
```diff
@@ -85,7 +85,11 @@ class _MyAppState extends State<MyApp> {
   // Platform messages are asynchronous, so we initialize in an async method.
   Future<void> initPlatformState() async {
     try {
-      datastorePlugin = AmplifyDataStore(modelProvider: ModelProvider.instance);
+      datastorePlugin = AmplifyDataStore(
+        modelProvider: ModelProvider.instance,
+        errorHandler: ((error) =>
+            {print("Custom ErrorHandler result: " + error.toString())}),
```
Comment (HuiSF) [inline packages/amplify_datastore/example/lib/main.dart:91]: mega nit: "result" sounds like we are expecting some thing to be returned intentionally, could be "Custom ErrorHandler received error" :D

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift [hunk 5]
```diff
@@ -121,10 +121,30 @@ public class SwiftAmplifyDataStorePlugin: NSObject, FlutterPlugin {
 
             flutterModelRegistration.version = modelProviderVersion
             let syncExpressions: [DataStoreSyncExpression] = try createSyncExpressions(syncExpressionList: syncExpressionList)
-
+            
             self.dataStoreHubEventStreamHandler?.registerModelsForHub(flutterModelRegistration: flutterModelRegistration)
+            
+
+            var errorHandler: DataStoreErrorHandler
+            if( (args["hasErrorHandler"] as? Bool) == true ){
```
Comment (HuiSF) [inline packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:129]:
```suggestion
            if((args["hasErrorHandler"] as? Bool) == true) {
```

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 6]
```diff
@@ -163,6 +165,18 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
 
         val dataStoreConfigurationBuilder = DataStoreConfiguration.builder()
 
+        if( (request["hasErrorHandler"] as? Boolean? == true) ) {
```
Comment (HuiSF) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:168]: I like the idea that defining a default error handler like what you did in this PR for the iOS implementation, can we adapt it to the Android implementation as well?

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift [hunk 7]
```diff
@@ -121,10 +121,30 @@ public class SwiftAmplifyDataStorePlugin: NSObject, FlutterPlugin {
...
+                errorHandler = { error in
+                    let map : [String:Any] = [
+                        "errorCode" : "DataStoreException",
+                        "errorMesage" : ErrorMessages.defaultFallbackErrorMessage,
+                        "details" : FlutterDataStoreErrorHandler.createSerializedError(error: error)
+                    ]
+                    self.channel!.invokeMethod("errorHandler", arguments: args)
+                }
+            }
+            else{
```
Comment (HuiSF) [inline packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:139]:
```suggestion
            } else {
```

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift [hunk 8]
```diff
@@ -133,7 +153,7 @@ public class SwiftAmplifyDataStorePlugin: NSObject, FlutterPlugin {
             
             Amplify.Logging.logLevel = .info
             print("Amplify configured with DataStore plugin")
-            result(true)
+            result(true)
```
Comment (HuiSF) [inline packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:156]:
```suggestion
            result(true)
```

File: packages/amplify_datastore/lib/method_channel_datastore.dart [hunk 9]
```diff
@@ -231,6 +252,23 @@ class AmplifyDataStoreMethodChannel extends AmplifyDataStore {
     return serializedItem["modelName"] as String;
   }
 
+  AmplifyException _deserializeExceptionFromMap(Map<String, dynamic> e) {
+    if (e['errorCode'] == 'DataStoreException') {
+      return DataStoreException.fromMap(Map<String, String>.from(e['details']));
+    } else if (e['errorCode'] == 'AmplifyAlreadyConfiguredException') {
+      return AmplifyAlreadyConfiguredException.fromMap(
+          Map<String, String>.from(e['details']));
+    } else {
```
Comment (HuiSF) [inline packages/amplify_datastore/lib/method_channel_datastore.dart:261]: It looks like the `else` covers use case that `e['errorCode'].isEmpty !== true`, with which there is an error code, but not handled by this `_deserializeExceptionFromMap` method, for this case categorizing them as "unhandled" could be more accurate than "missingExceptionMessage"?

File: packages/amplify_datastore_plugin_interface/lib/amplify_datastore_plugin_interface.dart [hunk 10]
```diff
@@ -80,15 +84,18 @@ abstract class DataStorePluginInterface extends AmplifyPluginInterface {
   }
 
   /// Configure AmplifyDataStore plugin with mandatory [modelProvider]
-  /// and optional datastore configuration properties including
+  /// and optional DataStore configuration properties including
+  ///
+  /// [errorHandler]: custom error handler code executed when DataStore encounters an unhandled error during its background operations
```
Comment (HuiSF) [inline packages/amplify_datastore_plugin_interface/lib/amplify_datastore_plugin_interface.dart:89]: Maybe this could be a bit clearer that we are expecting from the error handler. 

```suggestion
  /// [errorHandler]: Custom error handler function that may receive an [AmplifyException] object when DataStore encounters an unhandled error during its background operations
```

**Review by HuiSF — COMMENTED** [review]
Thanks for adding the error handler support to DataStore.

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt (hunk 2 above)
Comment (fjnoyp) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:18]: thanks - looks like it .. :)

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt (hunk 3 above)
Comment (fjnoyp) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:177]: thanks removed

File: packages/amplify_datastore/example/lib/main.dart (hunk 4 above)
Comment (fjnoyp) [reply packages/amplify_datastore/example/lib/main.dart:91]: nice good point!  updated

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift (hunk 5 above)
Comment (fjnoyp) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:129]: thanks updated

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt (hunk 6 above)
Comment (fjnoyp) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:168]: Good point !  Updated android: 

  ```
      var errorHandler : DataStoreErrorHandler;
        errorHandler = if( (request["hasErrorHandler"] as? Boolean? == true) ) {
            DataStoreErrorHandler {
                val args = hashMapOf(
                        "errorCode" to "DataStoreException",
                        "errorMessage" to ExceptionMessages.defaultFallbackExceptionMessage,
                        "details" to createSerializedError(it)
                )
                channel.invokeMethod("errorHandler", args)
            }
        }
        else {
            DataStoreErrorHandler {
                LOG.error(it.toString())
            }
        }
        dataStoreConfigurationBuilder.errorHandler(errorHandler)

```

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift (hunk 7 above)
Comment (fjnoyp) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:139]: thanks updated

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift (hunk 8 above)
Comment (fjnoyp) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:156]: my bad ... removed

File: packages/amplify_datastore/lib/method_channel_datastore.dart (hunk 9 above)
Comment (fjnoyp) [reply packages/amplify_datastore/lib/method_channel_datastore.dart:261]: Hmm makes sense.  However when looking at: 

```
  static const missingExceptionMessage =
      'An unknown exception has happened. ' + _createGithubIssueString;

```

It appears the existing error message returned is sufficient?  Do you think we should be saying "An unhandled exception has happened" instead? 

@HuiSF

File: packages/amplify_datastore_plugin_interface/lib/amplify_datastore_plugin_interface.dart (hunk 10 above)
Comment (fjnoyp) [reply packages/amplify_datastore_plugin_interface/lib/amplify_datastore_plugin_interface.dart:89]: Thanks updated with your suggestions!

**Review by ragingsquirrel3 — APPROVED** [review]

**Review by HuiSF — APPROVED** [review]
👍
